frontend/src/modules/entitieslist/actions.js

```diff
@@ -0,0 +1,65 @@
+/* @flow */
+
+export const REQUEST: 'entitieslist/REQUEST' = 'entitieslist/REQUEST';
+export const RECEIVE: 'entitieslist/RECEIVE' = 'entitieslist/RECEIVE';
```

Comment:
I was hoping Flow will make it possible to use union types for actions. Have you looked into it? In general, I don't see much value in binding action types to `consts`. Just using the actual string value seems enough. Would it be possible to do something like the following?

```js
type EntitiesListAction<T> = {
    type: "REQUEST" | "RECEIVE",
    data: T
};

function receive(
    entities: Array<Object>
): EntitiesListAction<Array<Object>> {
    return {
        type: "RECEIVE",
        data: entities,
    };
}
```

Reply:
There is one big advantage to using constants instead of strings: it prevent spelling mistakes. Also, since the app is going to be big, we want to have namespaced actions, but having to type the whole `namespace/ACTION_NAME` each time is quite tiring. Using constants makes this easier. 

On using unions with Flow, that's what we do indeed, see: https://github.com/mozilla/pontoon/pull/989/files/39f9c366cd9321704e5ae769424c8a4be2bc66dc#diff-ee54e21d7980bc484d37883b04204f69R7

I'm not a big fan of putting all data you pass to actions into a single sub-object (like `data` in your example, or in AMO they use `payload`). What do you think is the benefit of doing that?

Reply:
> There is one big advantage to using constants instead of strings: it prevent spelling mistakes.

Wouldn't Flow also prevent spelling mistakes in this case? The `type` field is defined as a `"REQUEST" | "RECEIVE"`; Flow should verify that only these values are used in the code.

> Also, since the app is going to be big, we want to have namespaced actions, but having to type the whole namespace/ACTION_NAME each time is quite tiring. Using constants makes this easier.

Good point about namespaces. I think my point about the union type of the `type` filed still stands though.

> I'm not a big fan of putting all data you pass to actions into a single sub-object (like data in your example, or in AMO they use payload). What do you think is the benefit of doing that?

The action object is more generic and unified and obeys the social contract: it always has a `type` field and a `data` (or `payload`) field.

This in turn means that you can use a generic for `data: T` and do away with specialized types per action, like `RequestAction` and `ReceiveAction`. They would be useful in a language which supports pattern matching on types of objects. But in JavaScript, you'll end up with the good old `switch (action.type)` anyways. To an extent, it defeats the purpose of having separate types.

Note that I haven't worked with Redux and Flow in a single project. All of the above is purely theoretical; I might be missing something here. I don't mean to block by any means. I'm just trying to help reduce the notorious boilerplate which Redux is often accused of adding to projects.

In the model I'm proposing, you only need one type of action per module, e.g. `EntitiesListAction`, whose type is an enum (or a union of string literals, really). I think this has a potential of reducing the boilerplate. If you prefer, I can try to prepare a PR demonstrating this approach after this PR lands. Again, I don't mean to block.

Reply:
@stasm I think what you're proposing doesn't work. What you get from doing what I currently do, `Action = Action1 | Action2`, is that it makes sure that for each action type, you have all the correct arguments. With a generic action `Action<T> = { type: "A" | "B", payload: T }` I believe (if I understand it correctly) that Flow will not be able to make sure that for type A you will have the correct type in the payload. 

Does that make sense?

Reply:
Hmm, that's exactly what the generic should take care of, though. Looking back at my original example:

```js
type EntitiesListAction<T> = {
    type: "REQUEST" | "RECEIVE",
    data: T
};

function receive(
    entities: Array<Object>
): EntitiesListAction<Array<Object>> {
    return {
        type: "RECEIVE",
        data: entities,
    };
}
```

You can only call `receive` with an array of objects as the argument, and the returned action will have an array of objects for `data`.

Reply:
Oh, I was talking about the reducer. It works for actions, but not for reducers: 

```javascript
function reducer(state: State, action: EntitiesListAction<any>): State {
    switch (action.type) {
        case "RECEIVE":
            // Flow won't be able to know what exactly `action.data` should contain for type = "RECEIVE"
            // Thus you can use attributes from type = "REQUEST" and it won't complain.
            // With my way of writing it, Flow will be able to tell that you can only use a specific
            // set of attributes and complain if you use anything else.
            console.log(action.data.entities);
    }
}
```

Reply:
I see. I didn't realize Flow checks the `switch` statement like you describe. That's cool!

Reply:
Also, https://flow.org/en/docs/react/redux/ is a good read and recommends an approach similar to yours.